docs/httpproxy.md

```diff
@@ -0,0 +1,833 @@
+spec:
+  virtualhost:
+    fqdn: bar1.bar.com
+  routes:
+    - conditions:
+        prefix: /
+      services:
+        - name: s2
+          port: 80
+```
+
+#### TLS
+
+HTTPProxy follows a similar pattern to Ingress for configuring TLS credentials.
+
+You can secure a HTTPProxy by specifying a Secret that contains TLS private key and certificate information.
+Contour (via Envoy) uses the SNI TLS extension to handle this behavior.
+If multiple HTTPProxy's utilize the same Secret, the certificate must include the necessary Subject Authority Name (SAN) for each fqdn.
+
+Contour also follows a "secure first" approach.
+When TLS is enabled for a virtual host any request to the insecure port is redirected to the secure interface with a 301 redirect.
+Specific routes can be configured to override this behavior and handle insecure requests by enabling the `spec.routes.permitInsecure` parameter on a Route.
+
+The TLS secret must contain keys named tls.crt and tls.key that contain the certificate and private key to use for TLS, e.g.:
+
+```yaml
+# ingress-tls.secret.yaml
+apiVersion: v1
+data:
+  tls.crt: base64 encoded cert
+  tls.key: base64 encoded key
+kind: Secret
+metadata:
+  name: testsecret
+  namespace: default
+type: kubernetes.io/tls
+```
+
+The HTTPProxy can be configured to use this secret using `tls.secretName` property:
+
+```yaml
+# tls.httpproxy.yaml
+apiVersion: projectcontour.io/v1alpha1
+kind: HTTPProxy
+metadata:
+  name: tls-example
+  namespace: default
+spec:
+  virtualhost:
+    fqdn: foo2.bar.com
+    tls:
+      secretName: testsecret
+  routes:
+    - conditions:
+        prefix: /
+      services:
+        - name: s1
+          port: 80
+```
+
+If the `tls.secretName` property contains a slash, eg. `somenamespace/somesecret` then, subject to TLS Certificate Delegation, the TLS certificate will be read from `somesecret` in `somenamespace`.
+See TLS Certificate Delegation below for more information.
+
+The TLS **Minimum Protocol Version** a vhost should negotiate can be specified by setting the `spec.virtualhost.tls.minimumProtocolVersion`:
+  - 1.3
+  - 1.2
+  - 1.1 (Default)
+
+A HTTPProxy can be configured to permit insecure requests to specific Routes.
+In this example, any request to `foo2.bar.com/blog` will not receive a 301 redirect to HTTPS, but the `/` route will:
+
+```yaml
+apiVersion: projectcontour.io/v1alpha1
+kind: HTTPProxy
+metadata: 
+  name: tls-example-insecure
+  namespace: default
+spec: 
+  virtualhost:
+    fqdn: foo2.bar.com
+    tls:
+      secretName: testsecret
+  routes: 
+    - conditions:
+        prefix: /
+      services: 
+        - name: s1
+          port: 80
+    - conditions:
+        prefix: /blog
+      permitInsecure: true
+      services: 
+        - name: s2
+          port: 80
+```
+
+#### Upstream TLS
+
+A HTTPProxy can proxy to an upstream TLS connection by first annotating the upstream Kubernetes service with: `contour.heptio.com/upstream-protocol.tls: "443,https"`.
```

Comment:
Annotation name update?

Reply:
I'd actually like to get rid of this annotation. We already specify which service/port to use, just need to add a field to the spec to upgrade the connection. 

My idea previously was to remove the `name` field of `service` and put the protocol in there (https://github.com/projectcontour/contour/pull/1195/files#diff-fc42c2911ab7ef4fd20292f823da57ebR116). This removes the need for annotations on services as well as can remove the `TCPProxy` spec.